app/controllers/incomeSources/add/BusinessReportingMethodController.scala

```diff
@@ -189,68 +190,44 @@ class BusinessReportingMethodController @Inject()(val authenticate: Authenticati
-              Logger("application").info(s"[BusinessReportingMethodController][updateReportingMethod]: " +
-                s"Errors received when updating tax year specific reporting methods: $result1\n$result2")
-              Redirect(redirectErrorUrl)
-            case (_: UpdateIncomeSourceResponseModel, _: UpdateIncomeSourceResponseError) =>
-              Logger("application").info(s"[BusinessReportingMethodController][updateReportingMethod]: " +
-                s"Updated tax year specific reporting method: $result1")
-              Logger("application").info(s"[BusinessReportingMethodController][updateReportingMethod]: " +
-                s"Error received when updating tax year specific reporting method: $result2")
-              //TODO: redirect to a new error page based on 1 success, 1 error
-              Redirect(redirectErrorUrl)
-            case (_: UpdateIncomeSourceResponseError, _: UpdateIncomeSourceResponseModel) =>
-              Logger("application").info(s"[BusinessReportingMethodController][updateReportingMethod]: " +
-                s"Error received when updating tax year specific reporting method: $result2")
-              Logger("application").info(s"[BusinessReportingMethodController][updateReportingMethod]: " +
-                s"Updated tax year specific reporting method: $result1")
-              //TODO: redirect to a new error page based on 1 success, 1 error
-              Redirect(redirectErrorUrl)
-            case (_: UpdateIncomeSourceResponseModel, _: UpdateIncomeSourceResponseModel) =>
-              Logger("application").info(s"[BusinessReportingMethodController][updateReportingMethod]: " +
-                s"Updated tax year specific reporting methods: $result1\n$result2")
-              Redirect(redirectUrl)
-          }
-        case _ =>
-          Logger("application").error("[BusinessReportingMethodController][updateReportingMethod]: " +
-            "Unexpected response received when updating tax year specific reporting methods")
+      manageReportingMethodUpdateResponses(results, redirectUrl) match {
+        case Left(ex) => Logger("application").error(s"[ForeignPropertyReportingMethodController][updateReportingMethod]: " +
+          s"Unable to update tax year specific reporting method: ${ex.getMessage}")
           Redirect(redirectErrorUrl)
+        case Right(redirectCall) => redirectCall
       }
     }.recover {
       case ex: Exception =>
-        Logger("application").error(s"[BusinessReportingMethodController][updateReportingMethod]: " +
+        Logger("application").error(s"[ForeignPropertyReportingMethodController][updateReportingMethod]: " +
           s"Error updating tax year specific reporting method: ${ex.getMessage}")
         Redirect(redirectErrorUrl)
     }
   }
 
+  @tailrec
+  private def manageReportingMethodUpdateResponses(results: Seq[UpdateIncomeSourceResponse], redirectUrl: Call): Either[Throwable, Result] = {
```

Comment:
In most of the cases should be sufficient to use foldLeft instead of custom recursion as this is hard to support in the future;

Reply:
as I understand underlying logic, this should be fine to find first error if any (or find all errors); 
otherwise collect success result in the log ( foldLeft would work fine for this) and redirect with success

Reply:
Here is potential solution for this processing:
```

      val errors = results.collect {
        case e: UpdateIncomeSourceResponseError => e
      }
      val success = results.collect {
        case c: UpdateIncomeSourceResponseModel => c
      }
      (errors, success) match {
        case (es: Seq[UpdateIncomeSourceResponseError], ss: UpdateIncomeSourceResponseModel) if es.length == 0 && ss.length == 0=>
          Logger("application").info(s"[BusinessReportingMethodController][updateReportingMethod]: " +
            s"Updated tax year specific reporting method: ${es.mkString(" - ")}")
          Redirect(redirectUrl)
        case (es: Seq[UpdateIncomeSourceResponseError] , _) if es.length > 0 =>
          Logger("application").info(s"[BusinessReportingMethodController][updateReportingMethod]: " +
            s"Updated tax year specific reporting method: ${es.mkString(" - ")}")
          Redirect(redirectUrl)
           // TODO: add other cases here ...
      }
```